File: ios/Classes/Core/TrackingScheduler.swift
import Foundation
import UserNotifications
import BackgroundTasks

/**
 * Manages scheduled tracking based on time windows
 * Automatically starts/stops LocationTracker at configured times
 */
class TrackingScheduler {

    // MARK: - Constants
    private static let TAG = "TrackingScheduler"
    private static let PREFS_KEY_SCHEDULE_ENABLED = "polyfence_schedule_enabled"
    private static let PREFS_KEY_TIME_WINDOWS = "polyfence_time_windows"
    private static let PREFS_KEY_START_IMMEDIATELY = "polyfence_start_immediately"

    // MARK: - Singleton
    static let shared = TrackingScheduler()

    // MARK: - Types

    /// Represents a time of day
    struct TimeOfDay {
        let hour: Int
        let minute: Int

        func toMinutesFromMidnight() -> Int {
            return hour * 60 + minute
        }

        static func fromMap(_ map: [String: Any]?) -> TimeOfDay? {
            guard let map = map,
                  let hour = map["hour"] as? Int,
                  let minute = map["minute"] as? Int else {
                return nil
            }
            return TimeOfDay(hour: hour, minute: minute)
        }
    }

    /// Represents a time window when tracking should be active
    struct TimeWindow {
        let startTime: TimeOfDay
        let endTime: TimeOfDay
        let daysOfWeek: [Int] // 1=Monday, 7=Sunday, empty=all days

        static func fromMap(_ map: [String: Any]) -> TimeWindow? {
            guard let startTime = TimeOfDay.fromMap(map["startTime"] as? [String: Any]),
                  let endTime = TimeOfDay.fromMap(map["endTime"] as? [String: Any]) else {
                return nil
            }
            let daysOfWeek = (map["daysOfWeek"] as? [Int]) ?? []
            return TimeWindow(startTime: startTime, endTime: endTime, daysOfWeek: daysOfWeek)
        }
    }

    /// Schedule configuration
    struct ScheduleConfig {
        var enabled: Bool = false
        var timeWindows: [TimeWindow] = []
        var startImmediatelyIfInWindow: Bool = true

        static func fromMap(_ map: [String: Any]?) -> ScheduleConfig {
            guard let map = map else { return ScheduleConfig() }

            let enabled = map["enabled"] as? Bool ?? false
            let startImmediately = map["startImmediatelyIfInWindow"] as? Bool ?? true
            let windowsList = (map["timeWindows"] as? [[String: Any]])?.compactMap { TimeWindow.fromMap($0) } ?? []

            return ScheduleConfig(enabled: enabled, timeWindows: windowsList, startImmediatelyIfInWindow: startImmediately)
        }
    }

    // MARK: - Properties
    private var config = ScheduleConfig()
    private var checkTimer: Timer?
    private weak var locationTracker: LocationTracker?

    // MARK: - Initialization
    private init() {}

    // MARK: - Public Methods

    /**
     * Set the LocationTracker reference for starting/stopping tracking
     */
    func setLocationTracker(_ tracker: LocationTracker) {
        self.locationTracker = tracker
    }

    /**
     * Update schedule configuration
     */
    func updateConfig(_ configMap: [String: Any]?) {
        config = ScheduleConfig.fromMap(configMap)
        saveConfig()

        if config.enabled {
            NSLog("[\(Self.TAG)] Schedule enabled with \(config.timeWindows.count) time windows")

            // Check if we should start tracking now
            if config.startImmediatelyIfInWindow && isCurrentlyInScheduledWindow() {
                NSLog("[\(Self.TAG)] Currently in scheduled window - starting tracking")
                startTracking()
            } else if !isCurrentlyInScheduledWindow() {
                NSLog("[\(Self.TAG)] Not in scheduled window - stopping tracking")
                stopTracking()
            }

            // Start periodic check timer
            startCheckTimer()
        } else {
            NSLog("[\(Self.TAG)] Schedule disabled - stopping timer")
            stopCheckTimer()
        }
    }

    /**
     * Check if current time is within any scheduled window
     */
    func isCurrentlyInScheduledWindow() -> Bool {
        if !config.enabled || config.timeWindows.isEmpty {
            return true // No schedule = always active
        }

        let now = Date()
        let calendar = Calendar.current
        let currentMinutes = calendar.component(.hour, from: now) * 60 + calendar.component(.minute, from: now)
        let currentDayOfWeek = getDayOfWeekIso(now) // 1=Monday, 7=Sunday

        for window in config.timeWindows {
            // Check if today is an allowed day
            if !window.daysOfWeek.isEmpty && !window.daysOfWeek.contains(currentDayOfWeek) {
                continue
            }

            let startMinutes = window.startTime.toMinutesFromMidnight()
            let endMinutes = window.endTime.toMinutesFromMidnight()

            // Handle windows that span midnight
            let isInWindow: Bool
            if endMinutes > startMinutes {
                // Normal window (e.g., 09:00 - 17:00)
                isInWindow = currentMinutes >= startMinutes && currentMinutes < endMinutes
            } else {
                // Window spans midnight (e.g., 22:00 - 06:00)
                isInWindow = currentMinutes >= startMinutes || currentMinutes < endMinutes
            }

            if isInWindow {
                return true
            }
        }

        return false
    }

    /**
     * Check if scheduling is enabled
     */
    func isEnabled() -> Bool {
        return config.enabled
    }

    /**
     * Load configuration from UserDefaults (called on app launch)
     */
    func loadConfig() {
        let defaults = UserDefaults.standard

        let enabled = defaults.bool(forKey: Self.PREFS_KEY_SCHEDULE_ENABLED)
        let startImmediately = defaults.bool(forKey: Self.PREFS_KEY_START_IMMEDIATELY)
        let windowsJson = defaults.string(forKey: Self.PREFS_KEY_TIME_WINDOWS) ?? ""

        var timeWindows: [TimeWindow] = []
        if !windowsJson.isEmpty {
            let windowStrings = windowsJson.split(separator: ";")
            for windowStr in windowStrings {
                let parts = windowStr.split(separator: "|")
                guard parts.count >= 2 else { continue }

                let startParts = parts[0].split(separator: ",")
                let endParts = parts[1].split(separator: ",")
                guard startParts.count == 2, endParts.count == 2,
                      let startHour = Int(startParts[0]),
                      let startMinute = Int(startParts[1]),
                      let endHour = Int(endParts[0]),
                      let endMinute = Int(endParts[1]) else { continue }

                var days: [Int] = []
                if parts.count > 2 && !parts[2].isEmpty {
                    days = parts[2].split(separator: ",").compactMap { Int($0) }
                }

                let window = TimeWindow(
                    startTime: TimeOfDay(hour: startHour, minute: startMinute),
                    endTime: TimeOfDay(hour: endHour, minute: endMinute),
                    daysOfWeek: days
                )
                timeWindows.append(window)
            }
        }

        config = ScheduleConfig(enabled: enabled, timeWindows: timeWindows, startImmediatelyIfInWindow: startImmediately)

        if config.enabled {
            NSLog("[\(Self.TAG)] Loaded schedule config with \(config.timeWindows.count) windows")
            startCheckTimer()
        }
    }

    // MARK: - Private Methods

    /**
     * Start the periodic check timer
     */
    private func startCheckTimer() {
        stopCheckTimer()

        // Check every minute
        checkTimer = Timer.scheduledTimer(withTimeInterval: 60.0, repeats: true) { [weak self] _ in
            self?.checkSchedule()
        }
        RunLoop.main.add(checkTimer!, forMode: .common)

        NSLog("[\(Self.TAG)] Schedule check timer started")
    }

    /**
     * Stop the periodic check timer
     */
    private func stopCheckTimer() {
        checkTimer?.invalidate()
        checkTimer = nil
    }

    /**
     * Check current schedule and start/stop tracking accordingly
     */
    private func checkSchedule() {
        let shouldTrack = isCurrentlyInScheduledWindow()
        let isTracking = locationTracker != nil

        if shouldTrack && !isTracking {
            NSLog("[\(Self.TAG)] Entered scheduled window - starting tracking")
            startTracking()
        } else if !shouldTrack && isTracking {
            NSLog("[\(Self.TAG)] Left scheduled window - stopping tracking")
            stopTracking()
        }
    }

    /**
     * Start tracking via LocationTracker
     */
    private func startTracking() {
        DispatchQueue.main.async { [weak self] in
            self?.locationTracker?.startTracking()
        }
    }

    /**
     * Stop tracking via LocationTracker
     */
    private func stopTracking() {
        DispatchQueue.main.async { [weak self] in
            self?.locationTracker?.stopTracking()
        }
    }

    /**
     * Convert Date to ISO day of week (1=Monday, 7=Sunday)
     */
    private func getDayOfWeekIso(_ date: Date) -> Int {
        let calendar = Calendar.current
        let weekday = calendar.component(.weekday, from: date)
        // Calendar.weekday: 1=Sunday, 2=Monday, ..., 7=Saturday
        // Convert to ISO: 1=Monday, 7=Sunday
        switch weekday {
        case 1: return 7  // Sunday
        case 2: return 1  // Monday
        case 3: return 2  // Tuesday
        case 4: return 3  // Wednesday
        case 5: return 4  // Thursday
        case 6: return 5  // Friday
        case 7: return 6  // Saturday
        default: return 1
        }
    }

    /**
     * Save configuration to UserDefaults
     */
    private func saveConfig() {
        let defaults = UserDefaults.standard

        defaults.set(config.enabled, forKey: Self.PREFS_KEY_SCHEDULE_ENABLED)
        defaults.set(config.startImmediatelyIfInWindow, forKey: Self.PREFS_KEY_START_IMMEDIATELY)

        // Serialize time windows
        let windowsJson = config.timeWindows.map { window in
            let days = window.daysOfWeek.map { String($0) }.joined(separator: ",")
            return "\(window.startTime.hour),\(window.startTime.minute)|\(window.endTime.hour),\(window.endTime.minute)|\(days)"
        }.joined(separator: ";")
        defaults.set(windowsJson, forKey: Self.PREFS_KEY_TIME_WINDOWS)

        defaults.synchronize()
    }
}

File: android/src/main/kotlin/com/polyfence/polyfence/core/TrackingScheduler.kt
package com.polyfence.polyfence.core

import android.app.AlarmManager
import android.app.PendingIntent
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.os.Build
import android.util.Log
import java.util.Calendar

/**
 * Manages scheduled tracking based on time windows
 * Automatically starts/stops LocationTracker at configured times
 */
class TrackingScheduler(private val context: Context) {

    companion object {
        private const val TAG = "TrackingScheduler"
        private const val PREFS_NAME = "polyfence_schedule"
        private const val KEY_SCHEDULE_ENABLED = "schedule_enabled"
        private const val KEY_TIME_WINDOWS = "time_windows"
        private const val KEY_START_IMMEDIATELY = "start_immediately_if_in_window"

        const val ACTION_SCHEDULE_START = "com.polyfence.SCHEDULE_START"
        const val ACTION_SCHEDULE_STOP = "com.polyfence.SCHEDULE_STOP"

        private const val REQUEST_CODE_START = 1001
        private const val REQUEST_CODE_STOP = 1002
    }

    private val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    private val alarmManager = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager

    /**
     * Data class representing a time of day
     */
    data class TimeOfDay(val hour: Int, val minute: Int) {
        fun toMinutesFromMidnight(): Int = hour * 60 + minute

        companion object {
            fun fromMap(map: Map<String, Any>?): TimeOfDay? {
                if (map == null) return null
                val hour = (map["hour"] as? Number)?.toInt() ?: return null
                val minute = (map["minute"] as? Number)?.toInt() ?: return null
                return TimeOfDay(hour, minute)
            }
        }
    }

    /**
     * Data class representing a time window when tracking should be active
     */
    data class TimeWindow(
        val startTime: TimeOfDay,
        val endTime: TimeOfDay,
        val daysOfWeek: List<Int> = emptyList() // 1=Monday, 7=Sunday, empty=all days
    ) {
        companion object {
            fun fromMap(map: Map<String, Any>): TimeWindow? {
                val startTime = TimeOfDay.fromMap(map["startTime"] as? Map<String, Any>) ?: return null
                val endTime = TimeOfDay.fromMap(map["endTime"] as? Map<String, Any>) ?: return null
                val daysOfWeek = (map["daysOfWeek"] as? List<*>)?.mapNotNull { (it as? Number)?.toInt() } ?: emptyList()
                return TimeWindow(startTime, endTime, daysOfWeek)
            }
        }
    }

    /**
     * Schedule settings configuration
     */
    data class ScheduleConfig(
        val enabled: Boolean = false,
        val timeWindows: List<TimeWindow> = emptyList(),
        val startImmediatelyIfInWindow: Boolean = true
    ) {
        companion object {
            fun fromMap(map: Map<String, Any>?): ScheduleConfig {
                if (map == null) return ScheduleConfig()

                val enabled = map["enabled"] as? Boolean ?: false
                val startImmediately = map["startImmediatelyIfInWindow"] as? Boolean ?: true
                val windowsList = (map["timeWindows"] as? List<*>)?.mapNotNull { windowMap ->
                    TimeWindow.fromMap(windowMap as? Map<String, Any> ?: return@mapNotNull null)
                } ?: emptyList()

                return ScheduleConfig(enabled, windowsList, startImmediately)
            }
        }
    }

    private var config = ScheduleConfig()

    /**
     * Update schedule configuration
     * @param configMap Configuration map from Flutter
     */
    fun updateConfig(configMap: Map<String, Any>?) {
        config = ScheduleConfig.fromMap(configMap)
        saveConfig()

        if (config.enabled) {
            Log.d(TAG, "Schedule enabled with ${config.timeWindows.size} time windows")

            // Check if we should start tracking now
            if (config.startImmediatelyIfInWindow && isCurrentlyInScheduledWindow()) {
                Log.d(TAG, "Currently in scheduled window - starting tracking")
                startTracking()
            } else if (!isCurrentlyInScheduledWindow()) {
                Log.d(TAG, "Not in scheduled window - stopping tracking")
                stopTracking()
            }

            // Schedule next alarm
            scheduleNextAlarm()
        } else {
            Log.d(TAG, "Schedule disabled - cancelling alarms")
            cancelAlarms()
        }
    }

    /**
     * Check if current time is within any scheduled window
     */
    fun isCurrentlyInScheduledWindow(): Boolean {
        if (!config.enabled || config.timeWindows.isEmpty()) {
            return true // No schedule = always active
        }

        val now = Calendar.getInstance()
        val currentMinutes = now.get(Calendar.HOUR_OF_DAY) * 60 + now.get(Calendar.MINUTE)
        val currentDayOfWeek = getDayOfWeekIso(now) // 1=Monday, 7=Sunday

        for (window in config.timeWindows) {
            // Check if today is an allowed day
            if (window.daysOfWeek.isNotEmpty() && currentDayOfWeek !in window.daysOfWeek) {
                continue
            }

            val startMinutes = window.startTime.toMinutesFromMidnight()
            val endMinutes = window.endTime.toMinutesFromMidnight()

            // Handle windows that span midnight
            val isInWindow = if (endMinutes > startMinutes) {
                // Normal window (e.g., 09:00 - 17:00)
                currentMinutes in startMinutes until endMinutes
            } else {
                // Window spans midnight (e.g., 22:00 - 06:00)
                currentMinutes >= startMinutes || currentMinutes < endMinutes
            }

            if (isInWindow) {
                return true
            }
        }

        return false
    }

    /**
     * Get the next schedule event (start or stop) time
     * Returns pair of (timestamp in millis, isStartEvent)
     */
    fun getNextScheduleEvent(): Pair<Long, Boolean>? {
        if (!config.enabled || config.timeWindows.isEmpty()) {
            return null
        }

        val now = Calendar.getInstance()
        val inWindow = isCurrentlyInScheduledWindow()

        var nextEventTime: Long = Long.MAX_VALUE
        var isStartEvent = !inWindow // If in window, next event is stop; if out, next is start

        for (window in config.timeWindows) {
            // Find next start time
            val nextStart = getNextOccurrence(window.startTime, window.daysOfWeek, now)
            // Find next end time
            val nextEnd = getNextOccurrence(window.endTime, window.daysOfWeek, now)

            if (inWindow && nextEnd < nextEventTime) {
                nextEventTime = nextEnd
                isStartEvent = false
            }

            if (!inWindow && nextStart < nextEventTime) {
                nextEventTime = nextStart
                isStartEvent = true
            }
        }

        return if (nextEventTime == Long.MAX_VALUE) null else Pair(nextEventTime, isStartEvent)
    }

    /**
     * Schedule the next alarm for start/stop
     */
    private fun scheduleNextAlarm() {
        cancelAlarms()

        val nextEvent = getNextScheduleEvent() ?: return
        val (nextTime, isStart) = nextEvent

        val action = if (isStart) ACTION_SCHEDULE_START else ACTION_SCHEDULE_STOP
        val requestCode = if (isStart) REQUEST_CODE_START else REQUEST_CODE_STOP

        val intent = Intent(context, ScheduleReceiver::class.java).apply {
            this.action = action
        }

        val pendingIntent = PendingIntent.getBroadcast(
            context,
            requestCode,
            intent,
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
        )

        // Use setExactAndAllowWhileIdle for reliable delivery
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            alarmManager.setExactAndAllowWhileIdle(AlarmManager.RTC_WAKEUP, nextTime, pendingIntent)
        } else {
            alarmManager.setExact(AlarmManager.RTC_WAKEUP, nextTime, pendingIntent)
        }

        val eventType = if (isStart) "START" else "STOP"
        val nextCal = Calendar.getInstance().apply { timeInMillis = nextTime }
        Log.d(TAG, "Scheduled $eventType alarm for ${nextCal.time}")
    }

    /**
     * Cancel all scheduled alarms
     */
    private fun cancelAlarms() {
        listOf(REQUEST_CODE_START, REQUEST_CODE_STOP).forEach { requestCode ->
            val action = if (requestCode == REQUEST_CODE_START) ACTION_SCHEDULE_START else ACTION_SCHEDULE_STOP
            val intent = Intent(context, ScheduleReceiver::class.java).apply {
                this.action = action
            }
            val pendingIntent = PendingIntent.getBroadcast(
                context,
                requestCode,
                intent,
                PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
            )
            alarmManager.cancel(pendingIntent)
        }
        Log.d(TAG, "Cancelled all schedule alarms")
    }

    /**
     * Get next occurrence of a time on allowed days
     */
    private fun getNextOccurrence(time: TimeOfDay, daysOfWeek: List<Int>, from: Calendar): Long {
        val cal = from.clone() as Calendar
        cal.set(Calendar.HOUR_OF_DAY, time.hour)
        cal.set(Calendar.MINUTE, time.minute)
        cal.set(Calendar.SECOND, 0)
        cal.set(Calendar.MILLISECOND, 0)

        // If time has already passed today, start checking from tomorrow
        if (cal.timeInMillis <= from.timeInMillis) {
            cal.add(Calendar.DAY_OF_YEAR, 1)
        }

        // If no specific days, return the calculated time
        if (daysOfWeek.isEmpty()) {
            return cal.timeInMillis
        }

        // Find next matching day
        for (i in 0 until 7) {
            val dayOfWeek = getDayOfWeekIso(cal)
            if (dayOfWeek in daysOfWeek) {
                return cal.timeInMillis
            }
            cal.add(Calendar.DAY_OF_YEAR, 1)
        }

        return cal.timeInMillis
    }

    /**
     * Convert Calendar day of week to ISO format (1=Monday, 7=Sunday)
     */
    private fun getDayOfWeekIso(cal: Calendar): Int {
        val calDay = cal.get(Calendar.DAY_OF_WEEK)
        return when (calDay) {
            Calendar.MONDAY -> 1
            Calendar.TUESDAY -> 2
            Calendar.WEDNESDAY -> 3
            Calendar.THURSDAY -> 4
            Calendar.FRIDAY -> 5
            Calendar.SATURDAY -> 6
            Calendar.SUNDAY -> 7
            else -> 1
        }
    }

    /**
     * Start tracking via LocationTracker service
     */
    private fun startTracking() {
        val intent = Intent(context, LocationTracker::class.java).apply {
            action = LocationTracker.ACTION_START_TRACKING
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            context.startForegroundService(intent)
        } else {
            context.startService(intent)
        }
    }

    /**
     * Stop tracking via LocationTracker service
     */
    private fun stopTracking() {
        val intent = Intent(context, LocationTracker::class.java).apply {
            action = LocationTracker.ACTION_STOP_TRACKING
        }
        context.startService(intent)
    }

    /**
     * Handle schedule alarm fired
     */
    fun handleAlarm(isStart: Boolean) {
        if (isStart) {
            Log.d(TAG, "Schedule START alarm fired - starting tracking")
            startTracking()
        } else {
            Log.d(TAG, "Schedule STOP alarm fired - stopping tracking")
            stopTracking()
        }

        // Schedule the next alarm
        scheduleNextAlarm()
    }

    /**
     * Save configuration to SharedPreferences
     */
    private fun saveConfig() {
        prefs.edit().apply {
            putBoolean(KEY_SCHEDULE_ENABLED, config.enabled)
            putBoolean(KEY_START_IMMEDIATELY, config.startImmediatelyIfInWindow)
            // Time windows serialized as JSON string for simplicity
            val windowsJson = config.timeWindows.joinToString(";") { window ->
                "${window.startTime.hour},${window.startTime.minute}|${window.endTime.hour},${window.endTime.minute}|${window.daysOfWeek.joinToString(",")}"
            }
            putString(KEY_TIME_WINDOWS, windowsJson)
            apply()
        }
    }

    /**
     * Load configuration from SharedPreferences
     */
    fun loadConfig() {
        val enabled = prefs.getBoolean(KEY_SCHEDULE_ENABLED, false)
        val startImmediately = prefs.getBoolean(KEY_START_IMMEDIATELY, true)
        val windowsJson = prefs.getString(KEY_TIME_WINDOWS, "") ?: ""

        val timeWindows = if (windowsJson.isNotEmpty()) {
            windowsJson.split(";").mapNotNull { windowStr ->
                try {
                    val parts = windowStr.split("|")
                    if (parts.size < 2) return@mapNotNull null

                    val startParts = parts[0].split(",")
                    val endParts = parts[1].split(",")
                    val days = if (parts.size > 2 && parts[2].isNotEmpty()) {
                        parts[2].split(",").mapNotNull { it.toIntOrNull() }
                    } else emptyList()

                    TimeWindow(
                        TimeOfDay(startParts[0].toInt(), startParts[1].toInt()),
                        TimeOfDay(endParts[0].toInt(), endParts[1].toInt()),
                        days
                    )
                } catch (e: Exception) {
                    Log.e(TAG, "Failed to parse time window: $windowStr", e)
                    null
                }
            }
        } else emptyList()

        config = ScheduleConfig(enabled, timeWindows, startImmediately)

        if (config.enabled) {
            Log.d(TAG, "Loaded schedule config with ${config.timeWindows.size} windows")
            scheduleNextAlarm()
        }
    }

    /**
     * Check if scheduling is enabled
     */
    fun isEnabled(): Boolean = config.enabled
}

/**
 * BroadcastReceiver for schedule alarms
 */
class ScheduleReceiver : BroadcastReceiver() {
    override fun onReceive(context: Context, intent: Intent) {
        val scheduler = TrackingScheduler(context)

        when (intent.action) {
            TrackingScheduler.ACTION_SCHEDULE_START -> {
                scheduler.handleAlarm(isStart = true)
            }
            TrackingScheduler.ACTION_SCHEDULE_STOP -> {
                scheduler.handleAlarm(isStart = false)
            }
            Intent.ACTION_BOOT_COMPLETED -> {
                // Restore schedule after device reboot
                scheduler.loadConfig()
            }
        }
    }
}
